File: go/driver/exec.go
// Command dispatch for the client-go driver.
package main

import (
	"context"
	"encoding/base64"
	"fmt"
	"os"
	"strings"

	"github.com/tikv/client-go/v2/oracle"
	"github.com/tikv/client-go/v2/tikv"
	"github.com/tikv/client-go/v2/txnkv"
	"github.com/tikv/client-go/v2/txnkv/transaction"
	"github.com/tikv/pd/client/opt"
)

// Command mirrors parity-proto's Rust enum (serde tag = "op", snake_case).
type Command struct {
	Op string `json:"op"`

	Name    string `json:"name,omitempty"`
	Session string `json:"session,omitempty"`
	Client  string `json:"client,omitempty"`
	Mode    string `json:"mode,omitempty"`

	Key     *KeyArg  `json:"key,omitempty"`
	Value   *KeyArg  `json:"value,omitempty"`
	Primary *KeyArg  `json:"primary,omitempty"`
	Keys    []KeyArg `json:"keys,omitempty"`

	Start *KeyArg `json:"start,omitempty"`
	End   *KeyArg `json:"end,omitempty"`
	Limit uint32  `json:"limit,omitempty"`
}

// KeyArg is {"s": "utf8"} or {"b64": "..."}.
type KeyArg struct {
	S   *string `json:"s,omitempty"`
	B64 *string `json:"b64,omitempty"`
}

func (k *KeyArg) Bytes() []byte {
	if k == nil {
		return nil
	}
	if k.S != nil {
		return []byte(*k.S)
	}
	if k.B64 != nil {
		raw, err := base64.StdEncoding.DecodeString(*k.B64)
		if err != nil {
			return nil
		}
		return raw
	}
	return nil
}

// Driver holds the sessions. One process per ROLE, never per language — client-go's
// config is a process global (config.GetGlobalConfig), so two differently-configured
// Go clients cannot cleanly coexist in one process. Making the process boundary the
// role boundary sidesteps that entirely.
type Driver struct {
	clients map[string]*txnkv.Client
	txns    map[string]*transaction.KVTxn
	// The committer a prewrite_only left mid-2PC, so `abandon` can drop it without
	// cleanup — that residue IS the state under test.
	committers map[string]*transaction.CommitterProbe
}

func NewDriver() *Driver {
	return &Driver{
		clients:    map[string]*txnkv.Client{},
		txns:       map[string]*transaction.KVTxn{},
		committers: map[string]*transaction.CommitterProbe{},
	}
}

func (d *Driver) Close() {
	for _, c := range d.clients {
		_ = c.Close()
	}
}

func pdAddrs() []string {
	v := os.Getenv("PD_ADDRS")
	if v == "" {
		v = "127.0.0.1:2379"
	}
	return strings.Split(v, ",")
}

// disableRouterClient turns OFF the PD "router client", and this is a DECLARED
// deviation from a stock client-go — see hello.config, which puts it in every trace.
//
// WHY IT IS NECESSARY. The pinned client-go (pins.toml client_go) depends on a PD
// client that, by default, resolves regions through the streaming `QueryRegion` RPC.
// The pinned cluster is PD v8.5.5 (pins.toml cluster — chosen because it is
// client-rust's own CI pin), and v8.5.5 does not implement that method:
//
//	rpc error: code = Unimplemented desc = unknown method QueryRegion for service pdpb.PD
//
// Left on, every region lookup fails and the driver hangs retrying forever. So the two
// pins are genuinely incompatible on this one PD API, and something has to give.
//
// WHY IT IS THE RIGHT THING TO GIVE. The router client is a region-lookup TRANSPORT
// optimization, not TiKV transaction semantics. Disabling it makes client-go resolve
// regions through the classic `GetRegion` RPC — which is the very path client-rust
// uses. For a harness whose whole purpose is to compare the two clients' behaviour,
// having them share a region-lookup path is more faithful, not less: it removes a
// difference that is about PD versions rather than about either client.
//
// WHY NOT JUST BUMP THE CLUSTER. Because the cluster pin is load-bearing elsewhere:
// the wyrd-gate's entire verdict is stated against v8.5.5, and bumping it would
// re-state every one of those claims. That is a deliberate, reviewed change, not
// something to do in passing to make a new driver start up.
//
// The honest cost: any gap that lives specifically IN the router path is invisible to
// this harness. That is a real limitation and it belongs in the ledger's scope notes,
// not buried here.
func disableRouterClient(cli *txnkv.Client) error {
	if err := cli.GetPDClient().UpdateOption(opt.EnableRouterClient, false); err != nil {
		return fmt.Errorf("cannot disable the PD router client: %w", err)
	}
	return nil
}

func (d *Driver) Execute(cmd Command) Response {
	ctx := context.Background()

	switch cmd.Op {
	case "hello":
		return Response{Hello: hello()}

	case "open_client":
		cli, err := txnkv.NewClient(pdAddrs())
		if err != nil {
			return Response{Observation: driverError(fmt.Sprintf("open_client: %v", err))}
		}
		if err := disableRouterClient(cli); err != nil {
			return Response{Observation: driverError(fmt.Sprintf("open_client: %v", err))}
		}
		d.clients[cmd.Name] = cli
		return Response{Observation: ok()}

	case "close_client":
		if cli, found := d.clients[cmd.Name]; found {
			_ = cli.Close()
			delete(d.clients, cmd.Name)
		}
		return Response{Observation: ok()}

	case "begin":
		cli, found := d.clients[cmd.Client]
		if !found {
			return Response{Observation: driverError("begin: no such client " + cmd.Client)}
		}
		txn, err := cli.Begin()
		if err != nil {
			return Response{Observation: driverError(fmt.Sprintf("begin: %v", err))}
		}
		// Go's default is optimistic; Rust's TransactionOptions::default() is
		// PESSIMISTIC. The scenario always states the mode explicitly so that
		// difference can never leak in as an unexamined default.
		if cmd.Mode == "pessimistic" {
			txn.SetPessimistic(true)
		}
		d.txns[cmd.Session] = txn
		return Response{Observation: ok().withStartTS(txn.StartTS())}

	case "put":
		txn, resp := d.txn(cmd.Session)
		if txn == nil {
			return resp
		}
		if err := txn.Set(cmd.Key.Bytes(), cmd.Value.Bytes()); err != nil {
			return Response{Observation: classify(err)}
		}
		return Response{Observation: ok()}

	case "get":
		txn, resp := d.txn(cmd.Session)
		if txn == nil {
			return resp
		}
		entry, err := txn.Get(ctx, cmd.Key.Bytes())
		if err != nil {
			return Response{Observation: classify(err)}
		}
		return Response{Observation: ok().withValue(entry.Value)}

	case "commit":
		txn, resp := d.txn(cmd.Session)
		if txn == nil {
			return resp
		}
		err := txn.Commit(ctx)
		delete(d.txns, cmd.Session)
		if err != nil {
			return Response{Observation: classify(err)}
		}
		return Response{Observation: ok().withCommitTS(txn.CommitTS())}

	case "rollback":
		txn, resp := d.txn(cmd.Session)
		if txn == nil {
			return resp
		}
		err := txn.Rollback()
		delete(d.txns, cmd.Session)
		if err != nil {
			return Response{Observation: classify(err)}
		}
		return Response{Observation: ok()}

	case "snapshot_get":
		// A read OUTSIDE any transaction, at a fresh timestamp. This is the parity
		// claim in the orphaned-lock scenario: "is the key readable again?"
		cli, found := d.clients[cmd.Client]
		if !found {
			return Response{Observation: driverError("snapshot_get: no such client " + cmd.Client)}
		}
		ts, err := cli.CurrentTimestamp(oracle.GlobalTxnScope)
		if err != nil {
			return Response{Observation: driverError(fmt.Sprintf("snapshot_get: ts: %v", err))}
		}
		snap := cli.GetSnapshot(ts)
		entry, err := snap.Get(ctx, cmd.Key.Bytes())
		if err != nil {
			return Response{Observation: classify(err)}
		}
		return Response{Observation: ok().withValue(entry.Value)}

	case "scan_locks":
		// Ground truth for durable lock residue, and symmetric with the Rust
		// driver's TransactionClient::scan_locks. This is what makes findings 1 and
		// 2 DIFFABLE rather than merely assertable.
		cli, found := d.clients[cmd.Client]
		if !found {
			return Response{Observation: driverError("scan_locks: no such client " + cmd.Client)}
		}
		ts, err := cli.CurrentTimestamp(oracle.GlobalTxnScope)
		if err != nil {
			return Response{Observation: driverError(fmt.Sprintf("scan_locks: ts: %v", err))}
		}
		probe := tikv.StoreProbe{KVStore: cli.KVStore}
		locks, err := probe.ScanLocks(ctx, cmd.Start.Bytes(), cmd.End.Bytes(), ts)
		if err != nil {
			return Response{Observation: classify(err)}
		}
		return Response{Observation: ok().withLocks(mkLocks(locks))}

	case "prewrite_only":
		return Response{Observation: d.prewriteOnly(ctx, cmd)}

	case "abandon":
		// Drop the txn WITHOUT commit or rollback — "the crash". The locks a
		// prewrite left behind stay behind, which is the entire point.
		delete(d.txns, cmd.Session)
		delete(d.committers, cmd.Session)
		return Response{Observation: ok()}

	default:
		return Response{Observation: driverError("unknown op: " + cmd.Op)}
	}
}

func (d *Driver) txn(session string) (*transaction.KVTxn, Response) {
	txn, found := d.txns[session]
	if !found {
		return nil, Response{Observation: driverError("no such session: " + session)}
	}
	return txn, Response{}
}

func (o *Observation) withValue(v []byte) *Observation {
	b := mkBytes(v)
	o.Value = &b
	return o
}

func (o *Observation) withLocks(l []Lock) *Observation {
	o.Locks = l
	return o
}

func (o *Observation) withStartTS(ts uint64) *Observation {
	o.StartTS = ts
	return o
}

func (o *Observation) withCommitTS(ts uint64) *Observation {
	o.CommitTS = ts
	return o
}

File: crates/rust-driver/src/exec.rs
//! Command dispatch for the client-rust driver.

use std::collections::HashMap;

use parity_proto::Class;
use parity_proto::Command;
use parity_proto::Hello;
use parity_proto::LockObs;
use parity_proto::Observation;
use parity_proto::TxnMode;
use parity_proto::PROTOCOL_VERSION;
use serde::Serialize;
use tikv_client::CheckLevel;
use tikv_client::Transaction;
use tikv_client::TransactionClient;
use tikv_client::TransactionOptions;
// `Timestamp::version()` lives on this trait, not the struct.
use tikv_client::TimestampExt;

use crate::mapping::classify;

/// One reply. Exactly one field is set.
#[derive(Debug, Serialize)]
pub struct Response {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub hello: Option<Hello>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub observation: Option<Observation>,
}

impl Response {
    pub fn observation(obs: Observation) -> Self {
        Self {
            hello: None,
            observation: Some(obs),
        }
    }
    fn hello(h: Hello) -> Self {
        Self {
            hello: Some(h),
            observation: None,
        }
    }
}

pub struct Driver {
    clients: HashMap<String, TransactionClient>,
    txns: HashMap<String, Transaction>,
}

impl Driver {
    pub fn new() -> Self {
        Self {
            clients: HashMap::new(),
            txns: HashMap::new(),
        }
    }

    pub async fn execute(&mut self, cmd: Command) -> Response {
        match cmd {
            Command::Hello => Response::hello(Hello {
                driver: "rust".to_owned(),
                protocol: PROTOCOL_VERSION.to_owned(),
                client: parity_proto::command::ClientId {
                    name: "tikv-client".to_owned(),
                    // A path dependency carries no version Cargo can report, so the
                    // driver cannot witness its own revision the way the Go driver
                    // can via ReadBuildInfo. That hole is closed OUT OF BAND, by
                    // scripts/provenance.sh recording the sibling checkout's git rev
                    // and refusing to run off-pin under PARITY_STRICT.
                    version: "path:../client-rust (see results/provenance.json)".to_owned(),
                    replaced: false,
                },
                features: vec!["scan_locks".to_owned()],
                // DECLARE THE DEVIATION. Dropping an uncommitted Transaction PANICS by
                // default (CheckLevel::Panic), which would kill the driver mid-scenario
                // and make `abandon` — "the crash" — unimplementable. So every txn is
                // begun with drop_check(Warn). That is a difference from a stock client
                // and it belongs in the trace, not in a code comment nobody reads six
                // months from now.
                config: vec!["drop_check=warn (stock default is panic)".to_owned()],
            }),

            Command::OpenClient { name } => {
                match TransactionClient::new(harness::pd_addrs()).await {
                    Ok(c) => {
                        self.clients.insert(name, c);
                        Response::observation(Observation::ok())
                    }
                    Err(e) => Response::observation(Observation::driver_error(format!(
                        "open_client: {e}"
                    ))),
                }
            }

            Command::CloseClient { name } => {
                self.clients.remove(&name);
                Response::observation(Observation::ok())
            }

            Command::Begin {
                session,
                client,
                mode,
            } => {
                let Some(c) = self.clients.get(&client) else {
                    return Response::observation(Observation::driver_error(format!(
                        "begin: no such client {client}"
                    )));
                };
                // Always state the mode explicitly. Rust's TransactionOptions::default()
                // is PESSIMISTIC while Go's Begin() is OPTIMISTIC — a difference that
                // must never leak in as an unexamined default and silently make the two
                // runs incomparable.
                let opts = match mode {
                    TxnMode::Optimistic => TransactionOptions::new_optimistic(),
                    TxnMode::Pessimistic => TransactionOptions::new_pessimistic(),
                }
                .drop_check(CheckLevel::Warn);

                match c.begin_with_options(opts).await {
                    Ok(txn) => {
                        let ts = txn.start_timestamp().version();
                        self.txns.insert(session, txn);
                        Response::observation(Observation::ok().with_start_ts(ts))
                    }
                    Err(e) => Response::observation(classify(&e)),
                }
            }

            Command::Put {
                session,
                key,
                value,
            } => match self.txns.get_mut(&session) {
                Some(txn) => match txn.put(key.bytes(), value.bytes()).await {
                    Ok(()) => Response::observation(Observation::ok()),
                    Err(e) => Response::observation(classify(&e)),
                },
                None => Response::observation(no_session(&session)),
            },

            Command::Get { session, key } => match self.txns.get_mut(&session) {
                Some(txn) => match txn.get(key.bytes()).await {
                    // Rust returns Ok(None); Go returns ErrNotExist. ONE FACT, two
                    // idioms — normalize the presentation.
                    Ok(None) => Response::observation(Observation::new(Class::NotFound)),
                    Ok(Some(v)) => Response::observation(Observation::ok().with_value(Some(v))),
                    Err(e) => Response::observation(classify(&e)),
                },
                None => Response::observation(no_session(&session)),
            },

            Command::Commit { session } => match self.txns.remove(&session) {
                Some(mut txn) => match txn.commit().await {
                    Ok(Some(ts)) => {
                        Response::observation(Observation::ok().with_commit_ts(ts.version()))
                    }
                    Ok(None) => Response::observation(Observation::ok()),
                    Err(e) => Response::observation(classify(&e)),
                },
                None => Response::observation(no_session(&session)),
            },

            Command::Rollback { session } => match self.txns.remove(&session) {
                Some(mut txn) => match txn.rollback().await {
                    Ok(()) => Response::observation(Observation::ok()),
                    Err(e) => Response::observation(classify(&e)),
                },
                None => Response::observation(no_session(&session)),
            },

            // A read OUTSIDE any transaction, at a fresh timestamp: "is the key
            // readable again?" — the parity claim in the orphaned-lock scenario.
            Command::SnapshotGet { client, key } => {
                let Some(c) = self.clients.get(&client) else {
                    return Response::observation(Observation::driver_error(format!(
                        "snapshot_get: no such client {client}"
                    )));
                };
                let ts = match c.current_timestamp().await {
                    Ok(ts) => ts,
                    Err(e) => {
                        return Response::observation(Observation::driver_error(format!(
                            "snapshot_get: current_timestamp: {e}"
                        )))
                    }
                };
                // `TransactionClient::snapshot` applies .read_only() internally
                // (client.rs:230) — a claim the harness once got WRONG and retracted.
                let mut snap = c.snapshot(ts, TransactionOptions::new_optimistic());
                match snap.get(key.bytes()).await {
                    Ok(None) => Response::observation(Observation::new(Class::NotFound)),
                    Ok(Some(v)) => Response::observation(Observation::ok().with_value(Some(v))),
                    Err(e) => Response::observation(classify(&e)),
                }
            }

            // Ground truth for durable lock residue, and it must answer the SAME QUESTION
            // as the Go driver's StoreProbe.ScanLocks: "every lock in [start, end)".
            //
            // The two clients do not offer that question at the same altitude, and the
            // difference is a trap. `TransactionClient::scan_locks`'s third argument is a
            // BATCH SIZE, not a limit: the plan (retry_multi_region -> Collect) returns up
            // to that many locks PER REGION and does not page within one. client-go's
            // probe pages to exhaustion (an internal loop at 1024/iteration). Pass a batch
            // size through naively and the two drivers silently answer different
            // questions — Rust truncating where Go does not — which would manufacture a
            // lock-count divergence out of harness semantics rather than client behaviour.
            //
            // So the Rust driver PAGES, advancing past the last key it saw, exactly as
            // client-go's own probe does. That is not papering over a client deficiency:
            // the batch API is doing precisely what it says, and both clients' test
            // helpers have to loop over it. What would be dishonest is truncating Go's
            // result to match Rust's cap — that would HIDE locks, and the whole point of
            // this observation is to see what residue is left behind.
            Command::ScanLocks {
                client,
                start,
                end,
                batch_size,
            } => {
                let Some(c) = self.clients.get(&client) else {
                    return Response::observation(Observation::driver_error(format!(
                        "scan_locks: no such client {client}"
                    )));
                };
                let ts = match c.current_timestamp().await {
                    Ok(ts) => ts,
                    Err(e) => {
                        return Response::observation(Observation::driver_error(format!(
                            "scan_locks: current_timestamp: {e}"
                        )))
                    }
                };

                let end_key = end.bytes();
                let mut cursor = start.bytes();
                let mut locks: Vec<LockObs> = Vec::new();

                loop {
                    let batch = match c
                        .scan_locks(&ts, cursor.clone()..end_key.clone(), batch_size)
                        .await
                    {
                        Ok(b) => b,
                        Err(e) => return Response::observation(classify(&e)),
                    };
                    if batch.is_empty() {
                        break;
                    }

                    // Advance past the greatest key we saw. `Collect` merges per-region
                    // results, so the batch is not guaranteed sorted; take the max rather
                    // than the last, or a page boundary could rewind the cursor and loop
                    // forever.
                    let Some(furthest) = batch.iter().map(|l| l.key.clone()).max() else {
                        break;
                    };

                    let before = locks.len();
                    locks.extend(batch.into_iter().map(|l| LockObs {
                        key: l.key.into(),
                        primary: l.primary_lock.into(),
                        kind: lock_kind(l.lock_type),
                        ttl_ms: l.lock_ttl,
                        txn_start_ts: l.lock_version,
                    }));

                    // A page that added nothing new cannot make progress, and the cursor
                    // is about to be set past `furthest` anyway; stop rather than spin.
                    if locks.len() == before {
                        break;
                    }

                    // Next key after `furthest`: append a 0 byte (no key sorts between
                    // `k` and `k\0`), which is the standard exclusive-successor trick and
                    // cannot overflow the way an increment could.
                    cursor = furthest;
                    cursor.push(0);
                    if cursor >= end_key {
                        break;
                    }
                }

                Response::observation(Observation::ok().with_locks(locks))
            }

            // ── THE ASYMMETRY, STATED RATHER THAN PAPERED OVER ───────────────
            // client-go drives 2PC one phase at a time with CommitterProbe, exported
            // from a NON-test file. client-rust's mocks are #[cfg(test)] and vanish
            // from the compiled crate, so there is no way to prewrite-and-stop through
            // its public API at all.
            //
            // The driver does NOT emulate this. Emulating it (say, by racing a second
            // txn to make the prewrite fail, as gate.rs::d6 must) would be exactly the
            // "workaround for a client deficiency" the repo's governing principle
            // forbids — and it would make the two runs incomparable, because the
            // SETUP would no longer be held constant.
            //
            // So it answers `unsupported`, which is a legitimate, comparable
            // observation. And the gap it names — client-rust exposes no importable
            // test probes — is itself a finding worth filing: it is *why* d6 needs a
            // region-split trick and d7 needs a compile-time failpoint to manufacture
            // states client-go's own tests simply construct.
            Command::PrewriteOnly { .. } => Response::observation(Observation::unsupported(
                "client-rust exposes no test probes (its mocks are #[cfg(test)]), so a \
                 caller cannot drive 2PC phase-by-phase. client-go exports CommitterProbe \
                 from a non-test file. The driver does NOT emulate this: the gap is the finding.",
            )),

            // Drop the txn WITHOUT commit or rollback — "the crash". Safe only because
            // Begin set drop_check(Warn); with the stock CheckLevel::Panic this would
            // abort the driver process.
            Command::Abandon { session } => {
                self.txns.remove(&session);
                Response::observation(Observation::ok())
            }
        }
    }
}

fn no_session(session: &str) -> Observation {
    Observation::driver_error(format!("no such session: {session}"))
}

/// Render `kvrpcpb::Op` BY NAME, matching the Go driver exactly. Both clients generate
/// this enum from the same proto, so the names line up; a name also survives a
/// renumbering that a bare integer would not.
fn lock_kind(op: i32) -> String {
    match op {
        0 => "put".to_owned(),
        1 => "del".to_owned(),
        2 => "lock".to_owned(),
        3 => "rollback".to_owned(),
        4 => "insert".to_owned(),
        5 => "pessimistic_lock".to_owned(),
        6 => "check_not_exists".to_owned(),
        other => format!("op_{other}"),
    }
}
